- Реализована логика создания стажировки с валидацией.

DmitryKirichenko opened the pull request:
- Реализована логика обновления стажировки.
- Добавлена логика получения всех стажировок.
- Добавлена логика получения стажировок с фильтрацией.
- Реализована логика получения стажировки по её ID.
- Реализована логика завершения стажировки и увольнения стажера.

DmitryKirichenko committed:
- Реализована логика создания стажировки с валидацией.
- Реализована логика обновления стажировки.
- Добавлена логика получения всех стажировок.
- Добавлена логика получения стажировок с фильтрацией.
- Реализована логика получения стажировки по её ID.
- Реализована логика завершения стажировки и увольнения стажера.

SergeyZhitenev commented on src/main/java/faang/school/projectservice/dto/InternshipDto.java:
```diff
@@ -0,0 +1,24 @@
+        Long id,
+        @NotBlank String name,
+        @NotNull Long projectId,
+        InternshipStatus status,
```
@NotNull

SergeyZhitenev commented on src/main/java/faang/school/projectservice/dto/InternshipDto.java:
```diff
@@ -0,0 +1,24 @@
+public record InternshipDto(
+        Long id,
+        @NotBlank String name,
+        @NotNull Long projectId,
```
Везде, где указывается id, добавь @Positive сейчас у тебя можно передать отрицательный id

SergeyZhitenev commented on src/main/java/faang/school/projectservice/mapper/InternshipMapper.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Named("mapLongToTeamMember")
+    default TeamMember mapLongToTeamMember(Long teamMemberId) {
+        return TeamMember.builder().id(teamMemberId).build();
```
TeamMember нужно не создавать, а получать из БД

SergeyZhitenev commented on src/main/java/faang/school/projectservice/mapper/InternshipMapper.java:
```diff
@@ -0,0 +1,35 @@
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface InternshipMapper {
+
+    @Mapping(source = "mentorId", target = "mentorId", qualifiedByName = "mapTeamMemberToLong")
```
Можно заменить на @Mapping(source = "team_member.id", target = "mentorId") и попробовать, должно работать

SergeyZhitenev commented on src/main/java/faang/school/projectservice/mapper/InternshipMapper.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.List;
+
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface InternshipMapper {
```
Ниже есть замечание по teamMember, но так же ты забыл про маппинг project, который так же должен вытаскиваться из бд, комментарии относительно него такие же. Скорее всего эти поля нужно заигнорить при маппинге и в сервис добавить метод инициализации, который будет ходить в бд и сетить эти поля вручную, а потом уже возвращать полный entity объект

SergeyZhitenev commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,80 @@
+    private final ProjectService projectService;
+    private final TeamMemberService teamMemberService;
+
+    public void createInternship(InternshipDto internshipDto) {
```
Замечание ко всем методам, каждый из них должен возвращать InternshipDto

SergeyZhitenev commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,80 @@
+        if (ChronoUnit.MONTHS.between(internshipDto.startDate(), internshipDto.endDate()) > 3) {
+            throw new DataValidationException("Продолжительность стажировки более 3х месяцев");
+        }
+        internshipRepository.save(internshipMapper.toEntity(internshipDto));
```
Не смаппится, почему описал в маппере. + нет проверок на существование проекта, переданного в дто, нет проверки на существование пользователей, переданных в дто. Когда добавишь отдельный метод для маппинга, у тебя эти проверки реализуются сами собой, потому что будешь обращаться в бд

SergeyZhitenev replied:
+ проверка, что юзер, переданный в качестве ментора, находится в команде в указанном проекте и обладает ролью ментор

SergeyZhitenev commented on src/main/java/faang/school/projectservice/dto/InternshipDto.java:
```diff
@@ -0,0 +1,24 @@
+        InternshipStatus status,
+        @NotNull Long mentorId,
+        @NotNull List<Long> internIds,
+        @Past LocalDateTime startDate,
```
Начало стажировки в прошлом?

SergeyZhitenev commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,64 @@
+
+    @PostMapping("/{internshipId}/complete/{teamMemberId}")
+    @ResponseStatus(HttpStatus.OK)
+    public void earlyCompletionInternship(@RequestParam long internshipId, @RequestParam long teamMemberId) {
```
Можно передать отрицательное значение, @Positive выручит

SergeyZhitenev commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,64 @@
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/v1/internship")
+public class InternshipController {
```
Все методы должны возвращать дто, сейчас пустое тело ответа

SergeyZhitenev commented on src/main/java/faang/school/projectservice/service/ProjectService.java:
```diff
@@ -0,0 +1,20 @@
+    private final ProjectRepository projectRepository;
+
+    private ProjectStatus getProjectStatusById(long id) {
+        return projectRepository.getProjectById(id).getStatus();
```
Если честно, выглядит как какие-то костыли, сделай отдельный метод получения project, а уже все статусы это валидация - ее можно вынести в валидатор

SergeyZhitenev commented on src/main/java/faang/school/projectservice/service/TeamMemberService.java:
```diff
@@ -0,0 +1,28 @@
+        return teamMemberRepository.findById(id);
+    }
+
+    public void setTeamMembersRoleNonIntern(List<Long> teamMemberIds) {
```
ты в 1 методе и идешь в бд за мемберами и меняешь им статусы, может кто-то скажет иное, но для меня выглядит странно + вынос 1 строки в метод ниже, читаемость прям падает

SergeyZhitenev commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    public void updateInternship(InternshipDto internshipDto) {
+        if (projectService.isProjectComplete(internshipDto.projectId())) {
```
В методе обновления, ты проверяешь, завершена ли стажировка или нет. Но у тебя нет метода, который позволяет статус стажировки менять (у тебя нет метода обновления стажировки)

SergeyZhitenev commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    public void dismissTeamMember(long internshipId, long teamMemberId) {
+        removeTeamMembersFromInternship(internshipId, teamMemberId);
```
зачем здесь разбивка на 2 метода? У тебя все что делает dismissTeamMember, это вызывает другой метод, все

SergeyZhitenev commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    private Internship getInternshipById(long internshipId) {
+        Internship internship = internshipRepository.findById(internshipId).orElse(null);
```
Зачем делать orElse(null), что само по себе bad practice, затем внизу делать if на проверку этого самого null и выкидывание исключения. Если можно просто указать orElseThrow()?

SergeyZhitenev commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,80 @@
+        removeTeamMembersFromInternship(internshipId, teamMemberId);
+    }
+
+    public List<InternshipDto> getFilteredInternships(InternshipStatus internshipStatus) {
```
![image](https://github.com/user-attachments/assets/4850bbc6-c45b-4d1a-aec2-5711009c355a)
а фильтр по роли где?

SergeyZhitenev commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @GetMapping("/filter")
+    public List<InternshipDto> getFilteredInternships(@RequestParam InternshipStatus internshipStatus) {
```
А как ты этот параметр получишь, если не ожидаешь его в строке запроса? + в задаче есть еще 1 фильтр, что, если передадут его? Да и объект из запроса как будто получить не получится

SergeyZhitenev commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,80 @@
+        }
+    }
+
+    public void completeInternship(long internshipId, long teamMemberId) {
```
наверное нужно еще и саму стажировку завершить 
![image](https://github.com/user-attachments/assets/658744eb-8706-4253-9672-f187681086f7)

SergeyZhitenev commented:
Нет тестов, соответственно код не проверялся, сейчас он не рабочий. Делай тесты сразу, все равно без них никто ничего не заапрувит + не нужно было бы ждать ревью, чтобы получить больше половины комментов. Так же надо тестить через постман, через него хотя бы можно убедиться, работает что-то или нет. Логика обновления, как мне показалось, реализована не так, как просится в задании. 
Мой совет - посмотри задачи других ребят, и те, что на ревью висят, и те, что уже выполнены и смержены ( даже открывать ничего не надо, все уже в проекте ). Это нормальная практика делать по образу и подобию, при этом разбирая то, что сделано) Делаем-с) 
+ мастер локально смержи, конфликты убрать

DmitryKirichenko committed:
-Доработал InternshipMapper
-Доработал InternshipController
-Переделал логику в InternshipService
-Создал InternshipFilterDto
-Написал тесты

YaKokin commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,72 @@
+        return new ResponseEntity<>(HttpStatus.OK);
+    }
+
+    @PutMapping("/{internshipId/change/{status}")
```
здесь скобочку забыл в мапинге. из-за этого код не компилится. большая просьба перед отправкой кода на ревью хотя бы запустить свой код и убедиться в том, что он компилируется.

YaKokin commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,72 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/v1/internship")
```
давай ресурс назовем во множественном числе /api/v1/internships
чтобы соответствовать общему код стайлу принятому в команде

YaKokin commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,72 @@
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/v1/internship")
+public class InternshipController {
```
Если делаешь версионирование, в название контроллер добавляй название версии. например InternshipV1Controller

YaKokin commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,72 @@
+    private final InternshipService internshipService;
+
+    @PostMapping
+    public ResponseEntity<Long> createInternship(@Valid @RequestBody InternshipDto internshipDto) {
```
при создании ресурса мы обычно возвращаем клиенту только что созданный ресурс (а точнее информацию о нем)
Возвращать просто лонг такая себе идея. Если клиенту захочется получить какую-то доп информацию о ресурсе, который ты создал, ему придется делать повторный запрос на бэк и попросту гонять трафик. давай лучше создалим специальную дто для респонса InternshipResponseDto и положим туда поля, который будут полезны клиенту при получении ответа?

YaKokin replied:
давай не будем явно заворачивать в ResponseEntity и просто вернем значение напрямую. спринг сам автоматически завернет возвращаемое значение в ResponseEntity

YaKokin commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    @PutMapping
+    public ResponseEntity<Void> updateInternship(@Valid @RequestBody InternshipDto internshipDto) {
```
такое же замечание. здесь я хочу видеть дто респонса, которое несет в себе информацию об обновленной сущности

YaKokin commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,72 @@
+        return ResponseEntity.ok(internshipService.createInternship(internshipDto));
+    }
+
+    @PutMapping
```
операцию обновлении существущего ресурса мы всегда применяем к конкретному ресурсу с конкретным id. исходя из твоего ендпоинта мне совсем не понятно, какой ресурс мы обновляем

YaKokin commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,72 @@
+        return new ResponseEntity<>(HttpStatus.OK);
+    }
+
+    @PostMapping("/{internshipId}/dismissal/{teamMemberId}")
```
хорошо спроектировал, молодец

YaKokin commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,72 @@
+        return new ResponseEntity<>(HttpStatus.OK);
+    }
+
+    @PostMapping("/{internshipId}/complete/{teamMemberId}")
```
и здесь, красавчик

YaKokin replied:
хотя я бы поставил метод DELETE чтобы явно указать, что мы удаляем участника из ресурса intership

YaKokin commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,72 @@
+        return new ResponseEntity<>(HttpStatus.OK);
+    }
+
+    @PostMapping("/filter/{internshipFilterDto}")
```
зачем здесь PostMapping?
давай лучше GetMapping
мы все таки получаем ресурс.

зачем ты указываешь дто в пути ура? при этом в самом методе ты ожидаешь дто из тела. давай будем из тела получать дто.

и "filter" убери из ендпоинта. он избыточен

YaKokin commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,72 @@
+        return ResponseEntity.ok(internshipService.getFilteredInternships(internshipFilterDto));
+    }
+
+    @GetMapping
```
молодец

YaKokin commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,72 @@
+        return ResponseEntity.ok(internshipService.getAllInternships());
+    }
+
+    @GetMapping("/{internshipId}")
```
отлично

YaKokin commented on src/main/java/faang/school/projectservice/mapper/InternshipMapper.java:
```diff
@@ -0,0 +1,70 @@
+    @Mapping(target = "mentorId", source = "entity", qualifiedByName = "getTeamMemberId")
+    @Mapping(target = "projectId", source = "entity", qualifiedByName = "getProjectId")
+    @Mapping(target = "internIds", source = "entity", qualifiedByName = "getTeamMembersIds")
+    InternshipDto toDto(Internship entity);
```
указывай пожалуйста в нейминге в какое конкретное дто ты мапишь. если у тебя будет несколько Internship Dto
то не понятно будет в какую конкретно ты мапишь.
предлагаю пока написать toInternshipDto

YaKokin commented on src/main/java/faang/school/projectservice/mapper/InternshipMapper.java:
```diff
@@ -0,0 +1,70 @@
+                                @Context ProjectService projectService
+    );
+
+    List<InternshipDto> toDtos(List<Internship> entities);
```
тут тоже поправь название

YaKokin commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,93 @@
+    private final TeamMemberService teamMemberService;
+
+    public Long createInternship(InternshipDto internshipDto) {
+        if (ChronoUnit.MONTHS.between(internshipDto.startDate(), internshipDto.endDate()) > 3) {
```
давай в отдельный метод вынесем валидацию, чтобы не городить

YaKokin commented on src/main/java/faang/school/projectservice/mapper/InternshipMapper.java:
```diff
@@ -0,0 +1,70 @@
+
+    @Named("getTeamMember")
+    default TeamMember getTeamMember(InternshipDto dto, @Context TeamMemberService teamMemberService) {
+        return teamMemberService.getTeamMemberById(dto.mentorId());
```
маппер лучше делать тупым. зависимость от сервиса делает маппер менее гибким. его тяжелее использовать повторно. также ты нарушаешь принцип единой ответственности, потому что твой маппер помимо преобразования объектов еще и сервисом умеет работать.
тестировать такой код будет тяжелее, потому что нужно будет думать об инжекции сервиса.

YaKokin replied:
лучше убрать получение сущностей по id в сервис

YaKokin commented on src/main/java/faang/school/projectservice/mapper/InternshipMapper.java:
```diff
@@ -0,0 +1,70 @@
+
+    @Named("getProject")
+    default Project getProject(InternshipDto dto, @Context ProjectService projectService) {
+        return projectService.getProjectById(dto.projectId());
```
аналогичное замечание

YaKokin commented on src/main/java/faang/school/projectservice/mapper/InternshipMapper.java:
```diff
@@ -0,0 +1,70 @@
+
+    @Named("getTeamMembersList")
+    default List<TeamMember> getTeamMembersList(InternshipDto dto, @Context TeamMemberService teamMemberService) {
+        return teamMemberService.getAllTeamMembersByIds(dto.internIds());
```
аналогично

YaKokin commented on src/main/java/faang/school/projectservice/dto/InternshipDto.java:
```diff
@@ -0,0 +1,25 @@
+@Builder
+public record InternshipDto(
+        Long id,
+        @NotBlank String name,
```
я бы еще поле description добавил

DmitryKirichenko committed:
-Переделал InternshipMapper
-Поправил маршруты в InternshipV1Controller

DmitryKirichenko committed:
-Поправил тесты

DmitryKirichenko committed:
Merge remote-tracking branch 'origin/Wyvern-master-stream7' into feature-BJS2-40709

# Conflicts:
#	src/main/java/faang/school/projectservice/repository/TeamMemberRepository.java
#	src/main/java/faang/school/projectservice/service/ProjectService.java
#	src/main/java/faang/school/projectservice/service/TeamMemberService.java
#	src/test/java/faang/school/projectservice/service/ProjectServiceTest.java
#	src/test/java/faang/school/projectservice/service/TeamMemberServiceTest.java

DmitryKirichenko committed:
Исправлены конфликты

DmitryKirichenko committed:
Исправил конфликты

DmitryKirichenko committed:
Revert "Исправил конфликты"

This reverts commit bbced999490a9c7113228c56b08c927a30df643b.

ipasynkov-epam commented on src/main/java/faang/school/projectservice/mapper/InternshipMapper.java:
```diff
@@ -0,0 +1,42 @@
+    Internship toEntity(InternshipDto dto);
+
+    List<Internship> toEntities(List<InternshipDto> dtos,
+                                @Context TeamMemberService teamMemberService,
```
поясни, пожалуйста, зачем ты используешь здесь сервисы и аннотацию Context? Не уверен, что в маппинге нужны ссылки на сервисы

ipasynkov-epam commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,100 @@
+    public InternshipDto createInternship(InternshipDto internshipDto) {
+        validateInternship(internshipDto);
+
+        Internship internship = internshipMapper.toEntity(internshipDto);
```
возможно стоит переименовать метод toEntity в маппере, чтобы понимать куда идет маппинг. Например, назвать его toInternship. Как думаешь?

ipasynkov-epam commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,100 @@
+
+        Internship internship = internshipMapper.toEntity(internshipDto);
+        internship.setMentorId(teamMemberService.getTeamMember(internshipDto.mentorId()));
+        internship.setInterns(teamMemberService.getAllTeamMembersByIds(internshipDto.internIds()));
```
не нужно делать в вызове метода вызовы других методов - это трудно для дебага, для отслеживания ошибок и пр. 
это к строкам: 32, 33, 35

ipasynkov-epam commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    public List<InternshipDto> getAllInternships() {
+        return internshipMapper.toInternshipDtos(internshipRepository.findAll());
```
тут снова вложенные вызовы. Репозиторий на findAll всегда возвращает валидные для маппера значения?

ipasynkov-epam commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    public InternshipDto getInternshipDtoById(long internshipId) {
+        return internshipMapper.toInternshipDto(getInternshipById(internshipId));
```
переход в метод getInternshipById скорее всего лишний. Мы здесь точно так же можем сходить в репозиторий за Internship и потом сделать маппинг

ipasynkov-epam replied:
Увидел, что он у тебя и в других методах тоже используется).

ipasynkov-epam commented:
Есть вопросы по маппингу. +нужно решить мерж-конфликты

DmitryKirichenko committed:
Merge remote-tracking branch 'origin/Wyvern-master-stream7' into feature-BJS2-40709

# Conflicts:
#	src/main/java/faang/school/projectservice/service/ProjectService.java
#	src/test/java/faang/school/projectservice/service/ProjectServiceTest.java

DmitryKirichenko committed:
-Переименовал методы InternshipMapper
-Исправил конфликты

ipasynkov-epam commented:
Сейчас ПР идет в ветку, у которой в названии есть слово master. Такие мержи блокируются правилами. Поэтому, сделай, пожалуйста, ПР в другую ветку и я его смержу
